### PR TITLE
New criteria

### PR DESCRIPTION
**Pull request by francescopeloi**
Changes proposed in this pull request:
- Added `CommissionsCriterion` to total the commissions paid across closed positions and `CommissionsImpactPercentageCriterion` to express how much those costs eat into gross profit
- Added `MaxConsecutiveLossCriterion`, `MaxConsecutiveProfitCriterion`, `MaxPositionNetLossCriterion` and `MaxPositionNetProfitCriterion` to report the worst loss streaks, best win streaks, and extreme per-position outcomes in a record
- Added `InPositionPercentageCriterion` to calculate the percentage of the time that a strategy remains invested

- [x] added an entry with related ticket number(s) to the unreleased section of `CHANGES.md`

### CODE REVIEW

File: ta4j-core/src/main/java/org/ta4j/core/criteria/InPositionPercentageCriterion.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            return numFactory.zero();
+        }
+        var totalDuration = totalTradingDuration(series);
+        if (totalDuration == 0 || tradingRecord.getPositionCount() == 0) {
+            return numFactory.zero();
+        }
+        var positionDuration = tradingRecord.getPositions().stream().mapToLong(p -> positionDuration(series, p)).sum();
+        return numFactory.numOf(positionDuration)
+                .dividedBy(numFactory.numOf(totalDuration))
+                .multipliedBy(numFactory.hundred());
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/criteria/InPositionPercentageCriterion.java]: For consistency should we return this in "decimal format" (i.e. without the * 100)?

For context please see https://github.com/ta4j/ta4j/pull/1074 and https://github.com/ta4j/ta4j/pull/949

"Percentage format" may still exist in some areas of the library but we should converge on a single, consistent way of representing percentages.

File: ta4j-core/src/main/java/org/ta4j/core/criteria/commissions/CommissionsCriterion.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+     * Calculates the commission paid for a single position.
+     *
+     * @param series   the bar series used for number creation
+     * @param position the evaluated position
+     * @return the commission paid for the provided position or zero when it is open
+     * @since 0.19
+     */
+    @Override
+    public Num calculate(BarSeries series, Position position) {
+        if (!position.isClosed()) {
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/criteria/commissions/CommissionsCriterion.java]: Should this only be for closed positions? It's common for exchanges/brokers to exact their transaction fees on a per trade basis so a commission was paid out regardless of the position's status.

File: ta4j-core/src/main/java/org/ta4j/core/criteria/commissions/CommissionsImpactPercentageCriterion.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            return zero;
+        }
+        var gross = p.getGrossProfit().abs();
+        var comm = zero;
+        var model = entry.getCostModel();
+        var transactionCost = model.calculate(p);
+        if (transactionCost != null) {
+            comm = transactionCost.abs();
+        }
+        return gross.isZero() ? zero : comm.dividedBy(gross).multipliedBy(numFactory.hundred());
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/criteria/commissions/CommissionsImpactPercentageCriterion.java]: as above, let's converge on "decimal" representation (0-1)

File: ta4j-core/src/main/java/org/ta4j/core/criteria/pnl/MaxConsecutiveLossCriterion.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+ */
+public class MaxConsecutiveLossCriterion extends AbstractAnalysisCriterion {
+
+    /**
+     * Returns the loss of the position when it ends in negative territory.
+     *
+     * @param series   the bar series used for number creation
+     * @param position the evaluated position
+     * @return the loss of the position or zero when it is not a loss
+     * @since 0.19
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/criteria/pnl/MaxConsecutiveLossCriterion.java]: Clarifying the since usage requirement a bit, this is fine to leave as-is but for future reference it's not required to tag overridden functions at the implementation layer.

File: ta4j-core/src/main/java/org/ta4j/core/criteria/pnl/MaxConsecutiveLossCriterion.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        for (var position : tradingRecord.getPositions()) {
+            if (position.isClosed() && position.getProfit().isNegative()) {
+                current = current.plus(position.getProfit());
+            } else {
+                if (current.isLessThan(max)) {
+                    max = current;
+                }
+                current = zero;
+            }
+        }
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/criteria/pnl/MaxConsecutiveLossCriterion.java]:
```suggestion
    for (var position : tradingRecord.getPositions()) {
    Num loss = calculate(series, position); // dogfood the position-based calculate
    if (loss.isNegative()) {
        current = current.plus(loss);
    } else {
        if (current.isLessThan(max)) {
            max = current;
        }
        current = zero;
    }
}
```

File: ta4j-core/src/main/java/org/ta4j/core/criteria/pnl/MaxConsecutiveProfitCriterion.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+     * @param tradingRecord the trading record containing the positions to scan
+     * @return the best consecutive profit or zero when there are no gains
+     * @since 0.19
+     */
+    @Override
+    public Num calculate(BarSeries series, TradingRecord tradingRecord) {
+        var zero = series.numFactory().zero();
+        var current = zero;
+        var max = zero;
+        for (var position : tradingRecord.getPositions()) {
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/criteria/pnl/MaxConsecutiveProfitCriterion.java]: dogfood the position based calculate if possible

File: ta4j-core/src/main/java/org/ta4j/core/criteria/commissions/CommissionsCriterion.java (hunk 2 above)
Comment (francescopeloi) [reply ta4j-core/src/main/java/org/ta4j/core/criteria/commissions/CommissionsCriterion.java]: Make sense, I've checked the various transaction cost model and they all handle open positions too. I'll change this.
